search.py

```diff
@@ -0,0 +1,55 @@
+from abc import ABC, abstractmethod
```

Comment:
Add comments

Reply:
Will add before the final merge into main.